src/core/net/dnssd_server.cpp

```diff
@@ -921,6 +949,154 @@ bool Server::Response::QueryNameMatchesService(const Srp::Server::Service &aServ
 
 #endif // OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
 
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE || OPENTHREAD_CONFIG_DNSSD_DISCOVERY_PROXY_ENABLE
+
+void Server::ConstructSoaServerName(void)
+{
+    if (mSoaServerName[0] == kNullChar)
+    {
+        StringWriter writer(mSoaServerName, sizeof(mSoaServerName));
+
+        writer.Append("otDNS%s", Get<Mac::Mac>().GetExtAddress().ToString().AsCString());
+    }
+}
+
+Error Server::Response::ResolveSoaOrNsQuery(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(QueryNameIsForDomain(kDefaultDomainName));
+
+    mSection = kAnswerSection;
+
+    // Handle AAAA (or ANY) query for the server host name used for
+    // SOA/NS answer.
+
+    if (mQuestions.IsFor(kRrTypeAaaa) || mQuestions.IsFor(kRrTypeAny))
+    {
+        Name::Buffer serverFullName;
+
+        ConstructFullName(Get<Server>().mSoaServerName, serverFullName);
+
+        if (QueryNameMatches(serverFullName))
+        {
+            for (Ip6::Netif::UnicastAddress &unicastAddr : Get<ThreadNetif>().GetUnicastAddresses())
+            {
+                if (!unicastAddr.mValid || !unicastAddr.mPreferred || unicastAddr.GetAddress().IsLinkLocalUnicast() ||
+                    Get<Mle::Mle>().IsMeshLocalAddress(unicastAddr.GetAddress()))
+                {
+                    continue;
+                }
+
+                SuccessOrExit(error = AppendAaaaRecord(unicastAddr.GetAddress(), kServerAaaaTtl));
+            }
+
+            if (IsEmpty())
+            {
+                SuccessOrExit(error = AppendAaaaRecord(Get<Mle::Mle>().GetMeshLocalEid(), kServerAaaaTtl));
+            }
+
+            ExitNow();
+        }
+    }
+
+    // Handle SOA or NS (or ANY) queries.
+
+    if (mQuestions.IsFor(kRrTypeSoa) || mQuestions.IsFor(kRrTypeNs) || mQuestions.IsFor(kRrTypeAny))
+    {
+        if (!QueryNameMatches(kDefaultDomainName))
+        {
+            // The SOA and NS query should be for the domain name
+            // (`default.service.arpa`). If it is for a name under the
+            // default domain, e.g. `somehost.default.service.arpa`, we
+            // still provide an answer in the Additional Data section
+            // unless the query is for ANY record type.
+
+            VerifyOrExit(!mQuestions.IsFor(kRrTypeAny));
+            mSection = kAdditionalDataSection;
```

Comment:
Based on definitions in  RFC 6762, 8766, 9665 (draft-SRP) there are no subdomains of `default.service.arpa.` and of `local.`. And mDNS does not have NS/SOA records by design. Therefore the server here must answer negatively in this case, where `QueryNameMatches` is false. This is done with an NXDOMAIN (3) answer, 0 answer records, 0 additional records, and 1 authority section record --  the SOA record.

So this is independent of whether a SOA or NS record is being requested.

For the case where `QueryNameMatches` is true, I believe the code is correct.

Reply:
This is inconsistent with what was discussed in SPEC-1389. I specifically asked about this in [comment](https://threadgroup.atlassian.net/browse/SPEC-1389?focusedCommentId=66993) and Ted's mentioned SOA should be included in additional data section in [comment](https://threadgroup.atlassian.net/browse/SPEC-1389?focusedCommentId=66997)

@EskoDijk can you please help clarify this?

Reply:
This was (I believe) a typo in Ted's comment that I didn't spot. Implementations I know use the "authority" section; and also RFC 8765 Section 6.1 has much detail.

The NXDOMAIN is my mistake: if a DP is active, it uses "negative answer" which is a NoError with 0 answers. I've updated this in the SPEC-1389 mini-spec comment.

Reply:
Update this to include the record in authority section.